harness/determined/deploy/local/cluster_utils.py

```diff
@@ -177,56 +185,167 @@ def master_up(
-        master_config_path = Path(temp_path)
+        master_config_path = pathlib.Path(temp_path)
 
     # This is always true by now, but mypy needs help.
     assert master_config_path is not None
+    restart_policy = "unless-stopped" if autorestart else "no"
+    env = {
+        "DET_MASTER_HTTP_PORT": str(port),
+        "DET_DB_PASSWORD": db_password,
+        "DET_LOG_INFO": "info",
+    }
 
-    env["DET_MASTER_CONFIG"] = str(master_config_path.resolve())
+    # Kill existing master container if exists.
+    master_down(master_name=master_name, delete_db=delete_db, cluster_name=cluster_name)
+
+    # Create network.
+    client = docker_client()
+
+    # Start db and wait for healthcheck.
+    db_name = f"{cluster_name}_{DB_NAME}"
+    volume_name = f"{cluster_name}_{VOLUME_NAME}"
+
+    @contextlib.contextmanager
+    def defer_cleanup(fn: Callable, *args: Any) -> Generator:
+        """
+        Defer cleanup tasks for each resource if Exceptions are caught.
+        """
+        try:
+            yield
+        except Exception as ex:
+            fn(*args)
+            raise ex
+
+    with contextlib.ExitStack() as exit_stack:
+        # Create network used by DB and master.
+        print(f"Creating network {NETWORK_NAME}...")
+        exit_stack.enter_context(defer_cleanup(remove_network, NETWORK_NAME))
+        client.networks.create(name=NETWORK_NAME, attachable=True)
+
+        # Start up db.
+        exit_stack.enter_context(defer_cleanup(db_down, db_name, VOLUME_NAME, delete_db))
+        db_up(
+            name=db_name,
+            password=db_password,
+            network_name=NETWORK_NAME,
+            cluster_name=cluster_name,
+            volume_name=volume_name,
+        )
+
+        # Wait for db to reach a healthy state.
+        _wait_for_container(db_name, timeout=5)
+        exit_stack.pop_all()
+
+        # Start master instance.
+        print(f"Creating {master_name}...")
+        exit_stack.enter_context(defer_cleanup(master_down, master_name, delete_db, cluster_name))
+        volumes = [f"{os.path.abspath(master_config_path)}:/etc/determined/master.yaml"]
+        client.containers.run(
+            image=f"{image_repo_prefix}/determined-master:{version}",
+            environment=env,
+            init=True,
+            mounts=[],
+            volumes=volumes,
+            name=master_name,
+            detach=True,
+            labels={},
+            restart_policy={"Name": restart_policy},
+            device_requests=None,
+            ports={"8080": f"{port}"},
+        )
 
-    master_down(master_name, delete_db)
-    docker_compose(command, master_name, env)
-    _wait_for_master("localhost", port, cluster_name)
+        # Connect to the network separately to set alias.
+        network = client.networks.get(NETWORK_NAME)
+        network.connect(container=master_name, aliases=["determined-master"])
 
+        _wait_for_master("localhost", port, cluster_name)
+        exit_stack.pop_all()
```

Comment:
This is a nice pattern, but not a very common pattern, so I'd say it's worth a comment explaining why you're doing this.